File: src/modules/bluetooth/LefuScaleModule.js
import LefuScale from "@modules/lefu-scale";

class LefuScaleModule {
  constructor() {
    this.lefuScale = LefuScale;
    console.log("LEFU_SCALE IMPORT HERE: ", this.lefuScale);
  }

  async initializeScale(key, secret) {
    return await this.lefuScale.initializeSdk(key, secret);
  }

  async startScan() {
    return this.lefuScale.startScan();
  }

  async stopScan() {
    return this.lefuScale.stopScan();
  }

  addDeviceDiscoveredListener(callback) {
    return this.lefuScale.addListener("onDeviceDiscovered", callback);
  }

  addBleStateChangeListener(callback) {
    return this.lefuScale.addListener("onBleStateChange", callback);
  }

  async connectToDevice(deviceId, disconnectTimeoutMillis) {
    return this.lefuScale.connectToDevice(deviceId, disconnectTimeoutMillis);
  }

  async disconnect() {
    return this.lefuScale.disconnect();
  }

  // async hello() {
  // 	return this.lefuScale.hello()
  // }

  // async getValueWithCallback(callback) {
  // 	return this.lefuScale.getValueWithCallback(callback)
  // }

  addWeightListener(callback) {
    return this.lefuScale.addListener("onWeightChange", callback);
  }

  addDisconnectListener(callback) {
    return this.lefuScale.addListener("hasDisconnected", callback);
  }

  // addConnectionStateListener(callback) {
  // 	return this.lefuScale.addListener('onConnectionStateChange', callback)
  // }

  // addErrorListener(callback) {
  // 	return this.lefuScale.addListener('onError', callback)
  // }
}

export default new LefuScaleModule();

File: modules/lefu-scale/android/src/main/java/expo/modules/lefuscale/LefuScaleModule.kt
package expo.modules.lefuscale

import expo.modules.kotlin.modules.Module
import com.lefu.ppbase.PPDeviceModel
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.lefuscale.device.LefuScaleService
import kotlinx.coroutines.*
import android.util.Log

class LefuScaleModule : Module() {
  private val moduleScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

  private var lefuService: LefuScaleService? = null

  override fun definition() = ModuleDefinition {
    Name("LefuScale")

    // Declare all events that can be sent to JavaScript.
    Events("onDeviceDiscovered", "onBleStateChange", "onWeightChange", "onConnectError")

    AsyncFunction("initializeSdk") { apiKey: String, apiSecret: String ->
      val context = requireNotNull(appContext.reactContext?.applicationContext) {
        "Application context is not available."
      }
      
      if (lefuService == null) {
        lefuService = LefuScaleService.instance
        setupLefuEventListeners()
      }
      lefuService?.initializeSdk(context, apiKey, apiSecret)
    }

    AsyncFunction("startScan") {
      lefuService?.startScan()
    }

    AsyncFunction("connectToDevice") { mac: String? ->
      mac?.let {
        lefuService?.connectToDevice(it)
      }
    }

    AsyncFunction("disconnect") {
      lefuService?.disconnect()
      sendEvent("onBleStateChange", mapOf("state" to "CustomPPBWorkSearchDeviceDisconnected"))
    }

    AsyncFunction("stopScan") {
      lefuService?.stopScan()
    }

    OnDestroy {
      // Cancel all coroutines when the module is destroyed to prevent memory leaks.
      moduleScope.cancel()
    }
  }

  private fun setupLefuEventListeners() {
    lefuService?.onDeviceDiscovered = { device ->
      val deviceInfo = mapOf(
        "name" to device.deviceName,
        "id" to device.deviceMac,
        "rssi" to device.rssi,
        "deviceType" to device.getDevicePeripheralType().name
      )
      sendEvent("onDeviceDiscovered", deviceInfo)
    }

    lefuService?.onBleStateChange = { state, _ ->
      if (lefuService?.deviceImpl?.lefuDevice === null){
        sendEvent("onBleStateChange", mapOf("state" to state.name))
      }
    }

    lefuService?.onConnectError = { errorMessage ->
      sendEvent("onConnectError", errorMessage)
    }

    lefuService?.onConnectionStateChange = { state ->
      sendEvent("onBleStateChange", mapOf("state" to state))
    }

    lefuService?.onWeightDataChange = { payload ->
      sendEvent("onWeightChange", payload)
    }
  }
}
